soa-derive-internal/src/vec.rs

```diff
@@ -71,97 +88,144 @@ pub fn derive(input: &Input) -> TokenStream {
+                        self.len -= 1;
+
+                        #(ptr::drop_in_place(self.data.#fields_names_1.ptr().add(i));)*
+                    }
+                }
             }
 
             /// Similar to [`
             #[doc = #vec_name_str]
             /// ::push()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.push).
             pub fn push(&mut self, value: #name) {
+                fn write_to_raw_vec<T>(buf: &mut RawVec<T>, value: T, index: usize) {
+                    unsafe {
+                        let ptr = buf.ptr().add(index);
+                        ptr::write(ptr, value);
+                    }
+                }
+
                 let #name{#(#fields_names_1),*} = value;
-                #(self.#fields_names_1.push(#fields_names_2);)*
+                self.reserve(1);
+                #(write_to_raw_vec(&mut self.data.#fields_names_1, #fields_names_2, self.len);)*
+                self.len += 1;
             }
 
             /// Similar to [`
             #[doc = #vec_name_str]
             /// ::len()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.len),
             /// all the fields should have the same length.
             pub fn len(&self) -> usize {
-                let len = self.#first_field.len();
-                #(debug_assert_eq!(self.#fields_names_1.len(), len);)*
-                len
+                self.len
             }
 
             /// Similar to [`
             #[doc = #vec_name_str]
             /// ::is_empty()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.is_empty),
             /// all the fields should have the same length.
             pub fn is_empty(&self) -> bool {
-                let empty = self.#first_field.is_empty();
-                #(debug_assert_eq!(self.#fields_names_1.is_empty(), empty);)*
-                empty
+                self.len == 0
             }
 
             /// Similar to [`
             #[doc = #vec_name_str]
             /// ::swap_remove()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.swap_remove).
             pub fn swap_remove(&mut self, index: usize) -> #name {
-                #(
-                    let #fields_names_1 = self.#fields_names_2.swap_remove(index);
-                )*
-                #name{#(#fields_names_1: #fields_names_2),*}
+                let length = self.len;
+                let slices = self.as_mut_slice();
+                #(slices.#fields_names_1.swap(index, length - 1);)*
+                self.pop().unwrap()
             }
 
             /// Similar to [`
             #[doc = #vec_name_str]
             /// ::insert()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.insert).
             pub fn insert(&mut self, index: usize, element: #name) {
+                fn insert_into_raw_vec<T>(buf: &mut RawVec<T>, len: usize, value: T, index: usize) {
+                    unsafe {
+                        // infallible
```

Comment:
Could you expand on what you mean here?

Reply:
This is from [the original std code](https://github.com/rust-lang/rust/blob/356d8ad1a310a2265e05d5c193d41704030d70d3/library/alloc/src/vec.rs#L1028).

I believe that what is saying is that this code is actually safe. Not 100% sure though.